File: leetcode/exercises/2.1.1.easy.valid_palindrome_1.go
// -----------------------
// Created     :3/11/2025
// Last Edited :3/11/2025 
// Topics      :
// Big O       :
// Problem Id  :25
// Source      :
// Notes       :
// -----------------------

package main

import (
    "fmt"
)

func isPalindrome(s string) bool {

    isalnum := func(c byte) bool {
        return (97 <= c && c <= 122) ||
               (65 <= c && c <= 90) ||
               (48 <= c && c <= 57)
    }

    bytetolower := func(c byte) byte {
        if 65 <= c && c <= 90 { 
            return c + 97 - 65
        } else {
            return c
        }
    }

    l := 0;
    r := len(s) - 1;
    for l < r {
        for !isalnum(s[l]) {
            l++
            if l == len(s) { return true }
        }
        for !isalnum(s[r]) {
            r--
        }

        if bytetolower(s[r]) != bytetolower(s[l]) { return false }
        l++
        r--
    }
    return true
}

func main() {
    fmt.Println(isPalindrome("A man, a plan, a canal: Panama"))
}

File: leetcode/exercises/2.1.1.easy.valid_palindrome_1.py
# -----------------------
# Created     : 12/11/2025
# Last Edited : 12/11/2025 
# Topics      : 
# Big O       :
# Problem Id  : 125
#  Source      :
#Notes       :
# -----------------------


class Solution:
    def isPalindrome(self, s: str) -> bool:
        if len(s) == 1 and s.isalpha(): return True

        left = 0
        right = len(s) - 1
        
        while left < right:

            while not s[left].isalnum():
                left += 1 
                if left == len(s): return True

            while not s[right].isalnum():
                right -= 1

            if s[left].lower() != s[right].lower(): 
                return False

            left += 1
            right -= 1

        return True

s = Solution()
print(s.isPalindrome("A man, a plan, a canal: Panama"))
print(s.isPalindrome(".,"))




        
